File: ecommerce_analysis.R
setwd('C:/Users/Abhinav/Downloads/My Assigments/Customer Analytics/Ecommerce_analysis/Ecommerce_analysis')

ecomdata <- read.csv('data.csv')

summary(ecomdata)

#converting Invoice date to Date format

ecomdata$InvoiceDate <- as.POSIXct(ecomdata$InvoiceDate,format = "%m/%d/%Y %H:%M")

ecomdata$CustomerID <- as.factor(ecomdata$CustomerID)
library(lubridate)

ecomdata$Month <- month(ecomdata$InvoiceDate)

ecomdata$Day <- day(ecomdata$InvoiceDate)

ecomdata$Hour <- hour(ecomdata$InvoiceDate)

ecomdata$Year <- year(ecomdata$InvoiceDate)

ecomdata$Month <- factor(ecomdata$Month,levels = seq(1,12),labels = c('Jan','Feb','Mar','Apr','May','June','July','Aug','Sept','Oct','Nov','Dec'))


library(dplyr)

#Check data for Returns

negquantdata <- ecomdata %>% filter(Quantity< 0)

#negquantdata <- ecomdata[ecomdata$Quantity< 0,]

ecomdata <- ecomdata %>% mutate(Total = Quantity * UnitPrice)

#Summarize the sales by Country, Year, Month

Conttotalsales <- ecomdata %>% group_by(Country) %>% summarise(sales = sum(Total)) %>% arrange(desc(sales))

Contyearsales <- ecomdata %>% group_by(Country,Year) %>% summarise(sales = sum(Total))

Contyearmonthsales <- ecomdata %>% group_by(Country,Year,Month) %>% summarise(sales = sum(Total))

monthsales <- ecomdata %>% group_by(Month) %>% summarise(sales = sum(Total)) %>% arrange(desc(sales))

yearmonthsales <- ecomdata %>% group_by(Year,Month) %>% summarise(sales = sum(Total))

daysales <- ecomdata %>% group_by(Day) %>% summarise(sales = sum(Total)) %>% arrange(desc(sales))

hoursales <- ecomdata %>% group_by(Hour) %>% summarise(sales = sum(Total))

#Creating visualization 

library(ggplot2)

# Removing returns, negtive price & missing customer ID

segdata <- ecomdata %>% filter(Quantity >= 0,UnitPrice >= 0,!is.na(CustomerID))

#Prepraing data for Customer segmentation

custdata <- segdata %>% select(CustomerID,Description)

library(reshape2)

custdata$value <- 1

cust_data <- dcast(custdata,CustomerID ~ Description)

clusterdata <- cust_data[,-1]

#Using PCA to reduce Dimension

prin_comp <- prcomp(clusterdata, scale. = T)

names(prin_comp)

#outputs the mean of variables
prin_comp$center

#outputs the standard deviation of variables
prin_comp$scale

dim(prin_comp$x)

biplot(prin_comp, scale = 0)

#compute standard deviation of each principal component
std_dev <- prin_comp$sdev

#compute variance
pr_var <- std_dev^2

#check variance of first 10 components
pr_var[1:10]

#proportion of variance explained
prop_varex <- pr_var/sum(pr_var)

prop_varex[1:20]

#scree plot

plot(prop_varex, xlab = "Principal Component",
     ylab = "Proportion of Variance Explained",
     type = "b")

#cumulative scree plot

plot(cumsum(prop_varex), xlab = "Principal Component",
     ylab = "Cumulative Proportion of Variance Explained",
     type = "b")

#Based on plot consider only first 1000 PC

pcadata <- data.frame(CustomerID = cust_data$CustomerID,prin_comp$x[,1:1000])

# scaling ( ordering is same)

#pcadata.sc <- scale(pcadata[,-1],center=TRUE,scale=TRUE)

#Elbow Method for finding the optimal number of clusters
set.seed(123)
# Compute and plot wss for k = 2 to k = 15.
k.max <- 15
data <- pcadata
wss <- sapply(10:k.max, 
              function(k){kmeans(data, k, nstart=50,iter.max = 15 )$tot.withinss})
wss
plot(10:k.max, wss,
     type="b", pch = 19, frame = FALSE, 
     xlab="Number of clusters K",
     ylab="Total within-clusters sum of squares")


# optimal number of cluster

library(NbClust)
set.seed(1234)
noculs <- NbClust(pcadata,min.nc=10,max.nc=15,method="kmeans")
table(noculs$Best.n[1,])
barplot(table(noculs$Best.n[1,]), 
        xlab="Numer of Clusters", ylab="Number of Criteria",
        main="Number of Clusters Chosen by 26 Criteria")

library("factoextra")
fviz_nbclust(noculs)

# Elbow method
fviz_nbclust(pcadata, kmeans, method = "wss") +
  geom_vline(xintercept = 4, linetype = 2)+
  labs(subtitle = "Elbow method")
# Silhouette method
fviz_nbclust(pcadata, kmeans, method = "silhouette")+
  labs(subtitle = "Silhouette method")
# Gap statistic
# nboot = 50 to keep the function speedy. 
# recommended value: nboot= 500 for your analysis.
# Use verbose = FALSE to hide computing progression.
set.seed(123)
fviz_nbclust(pcadata, kmeans, nstart = 25,  method = "gap_stat", nboot = 50)+
  labs(subtitle = "Gap statistic method")

#Based on optimal 
set.seed(1234)
results <- kmeans(pcadata, 10)

# understanding the output
results$size #size of each cluster

strength <- results$betweenss/results$tot.withinss # betweenss=distance between cluster should be large
# tot.withinss=avg. distance within cluster should be small
# strength should be higher

library(klaR)

cluster.results <-kmodes(clusterdata,5 )

cluster.results$size

cust_data$Cluster <- cluster.results$cluster

#Let do further analysis for cluster 1

cluster1 <- cust_data %>% filter(Cluster == 1)

clust_data1 <- cluster1[,-1]

cluster1.results <- kmodes(cluster1[,-1],5,iter.max = 10)

cluster1.results$size

# Using K prototype for customer segmentation

library(clustMixType)

kclust <- kproto(clusterdata, k=5, lambda = .8, iter.max = 100, nstart = 1)

clprofiles(kclust, clusterdata)
names(kclust)
str(kclust)

#to check the size of the respective clusters
kclust$size
kclust$cluster


#Prepraing data for market basket analysis

# Removing returns, negtive price & missing customer ID

segdata <- ecomdata %>% filter(Quantity >= 0,UnitPrice >= 0,!is.na(CustomerID))

summary(segdata)

proddata <- aggregate(Description ~ InvoiceNo, data = segdata,paste,collapse = ",")

library(splitstackshape)
custdata[,2]

mbadata <- cSplit(proddata, 'Description', sep=",", type.convert=FALSE)

#Conducting Market Basket Analysis

mbadata$InvoiceNo <- NULL

write.csv(mbadata,"mbadata.csv",quote = F,row.names = FALSE)

library(arules)
library(arulesViz)

itemdata <- read.transactions("mbadata.csv",sep = ",")

summary(itemdata)

itemFrequencyPlot(itemdata, topN = 20)

itemFrequencyPlot(itemdata, support = 0.03)

# we can control no of items by specifiying minlen and maxlen

itemrules <- apriori(itemdata, parameter = list(support = 0.001, confidence = 0.9, maxlen = 3))

#inspecting top 10 rules

inspect(itemrules[1:10])

#Removing duplicate rules

itemrules <- itemrules[!is.redundant(itemrules)]


itemrules<-sort(itemrules, by="confidence", decreasing=TRUE)


itemrules_df <- as(itemrules, "data.frame")

#Creating a visualization of the rules

topRules <- itemrules[1:10]

plot(topRules)

plot(topRules, method="graph")

plot(toprules,method="graph",interactive=TRUE,shading=NA)

plot(topRules, method = "grouped")

File: ecommerce_analysis.py
# -*- coding: utf-8 -*-
"""
Created on Mon Jun 25 19:33:22 2018

@author: Abhinav
"""

import os
import pandas as pd
import numpy as np

os.chdir('C:/Users/Abhinav/Downloads/My Assigments/Customer Analytics/Ecommerce_analysis/Ecommerce_analysis')

ecomdata = pd.read_csv('data.csv',encoding='latin1')

ecomdata.info()

ecomdata.describe()

#Converting Invoice Date to date format

ecomdata['InvoiceDate'] = pd.to_datetime(ecomdata['InvoiceDate'],format = "%m/%d/%Y %H:%M")

ecomdata['CustomerID'] = ecomdata['CustomerID'].astype('category')
#extracting Month

ecomdata['Month'] = ecomdata['InvoiceDate'].dt.month

ecomdata['Year'] = ecomdata['InvoiceDate'].dt.year

ecomdata['Day'] = ecomdata['InvoiceDate'].dt.day

ecomdata['Hour'] = ecomdata['InvoiceDate'].dt.hour

ecomdata['Month'] = ecomdata['Month'].astype('category').cat.rename_categories(['Jan','Feb','Mar','Apr','May','June','July','Aug','Sept','Oct','Nov','Dec'])

#Check data for returns

negquantdata = ecomdata[ecomdata['Quantity'] <0]

ecomdata['Total'] = ecomdata['Quantity'] * ecomdata['UnitPrice']

#Summarize the sales by Country, Year, Month

Conttotalsales = pd.DataFrame(ecomdata.groupby('Country')['Total'].sum()).reset_index()

Contyearsales = pd.DataFrame(ecomdata.groupby(['Country','Year'])['Total'].sum()).reset_index()

Contyearmonthsales = pd.DataFrame(ecomdata.groupby(['Country','Year','Month'])['Total'].sum()).reset_index()

monthsales = pd.DataFrame(ecomdata.groupby(['Month'])['Total'].sum()).reset_index()

yearmonthsales = pd.DataFrame(ecomdata.groupby(['Year','Month'])['Total'].sum()).reset_index()

daysales = pd.DataFrame(ecomdata.groupby(['Day'])['Total'].sum()).reset_index()

hoursales = pd.DataFrame(ecomdata.groupby(['Hour'])['Total'].sum()).reset_index()

# Removing returns, negtive price & missing customer ID
ecomdata['CustomerID'].isnull().values.any()

segdata = ecomdata[(ecomdata['Quantity'] >= 0) & (ecomdata['UnitPrice'] >= 0) & (pd.notnull(ecomdata['CustomerID']))]

#Prepraing data for Customer Segmentation

custdata = segdata[['CustomerID','Description']]

custdata['values'] = 1

cust_data = pd.crosstab(index = segdata['CustomerID'],columns =segdata['Description'])

#Using PCA to reduce dimension

from sklearn.decomposition import PCA

import matplotlib.pyplot as plt

from sklearn.preprocessing import scale

#convert it to numpy arrays
X= cust_data.values

#Scaling the values
X = scale(X)

pca = PCA(n_components=3877)

pca.fit(X)

#The amount of variance that each PC explains
var= pca.explained_variance_ratio_

#Cumulative Variance explains
var1=np.cumsum(np.round(pca.explained_variance_ratio_, decimals=4)*100)

plt.plot(var1)

#Looking at above plot I'm taking 30 variables
pca = PCA(n_components=1000)
pca.fit(X)
X1=pca.fit_transform(X)

from sklearn.cluster import KMeans
import matplotlib.cm as cm

cluster_errors = []


for num_clusters in range(10,16):
  clusters = KMeans( num_clusters )
  clusters.fit( X1 )
  cluster_errors.append( clusters.inertia_ )

clusters_df = pd.DataFrame( { "num_clusters":range(10,16), "cluster_errors": cluster_errors } )

plt.figure(figsize=(12,6))
plt.plot(clusters_df.num_clusters, clusters_df.cluster_errors, marker = "o" )

#Using Silhoutte method to obtain optimal no of cluster

from sklearn.metrics import silhouette_samples, silhouette_score

cluster_range = range( 10, 16 )

for n_clusters in cluster_range:
  # Create a subplot with 1 row and 2 columns
  fig, (ax1, ax2) = plt.subplots(1, 2)
  fig.set_size_inches(18, 7)

  # The 1st subplot is the silhouette plot
  # The silhouette coefficient can range from -1, 1 but in this example all
  # lie within [-0.1, 1]
  ax1.set_xlim([-0.1, 1])
  # The (n_clusters+1)*10 is for inserting blank space between silhouette
  # plots of individual clusters, to demarcate them clearly.
  ax1.set_ylim([0, len(X1) + (n_clusters + 1) * 10])

  # Initialize the clusterer with n_clusters value and a random generator
  # seed of 10 for reproducibility.
  clusterer = KMeans(n_clusters=n_clusters, random_state=10)
  cluster_labels = clusterer.fit_predict( X1 )

  # The silhouette_score gives the average value for all the samples.
  # This gives a perspective into the density and separation of the formed
  # clusters
  silhouette_avg = silhouette_score(X1, cluster_labels)
  print("For n_clusters =", n_clusters,
        "The average silhouette_score is :", silhouette_avg)

  # Compute the silhouette scores for each sample
  sample_silhouette_values = silhouette_samples(X1, cluster_labels)

  y_lower = 10
  for i in range(n_clusters):
      # Aggregate the silhouette scores for samples belonging to
      # cluster i, and sort them
      ith_cluster_silhouette_values = \
          sample_silhouette_values[cluster_labels == i]

      ith_cluster_silhouette_values.sort()

      size_cluster_i = ith_cluster_silhouette_values.shape[0]
      y_upper = y_lower + size_cluster_i

      color = cm.spectral(float(i) / n_clusters)
      ax1.fill_betweenx(np.arange(y_lower, y_upper),
                        0, ith_cluster_silhouette_values,
                        facecolor=color, edgecolor=color, alpha=0.7)

      # Label the silhouette plots with their cluster numbers at the middle
      ax1.text(-0.05, y_lower + 0.5 * size_cluster_i, str(i))

      # Compute the new y_lower for next plot
      y_lower = y_upper + 10  # 10 for the 0 samples

  ax1.set_title("The silhouette plot for the various clusters.")
  ax1.set_xlabel("The silhouette coefficient values")
  ax1.set_ylabel("Cluster label")

  # The vertical line for average silhoutte score of all the values
  ax1.axvline(x=silhouette_avg, color="red", linestyle="--")

  ax1.set_yticks([])  # Clear the yaxis labels / ticks
  ax1.set_xticks([-0.1, 0, 0.2, 0.4, 0.6, 0.8, 1])

  # 2nd Plot showing the actual clusters formed
  colors = cm.spectral(cluster_labels.astype(float) / n_clusters)
  ax2.scatter(X1[:, 0], X1[:, 1], marker='.', s=30, lw=0, alpha=0.7,
              c=colors)

  # Labeling the clusters
  centers = clusterer.cluster_centers_
  # Draw white circles at cluster centers
  ax2.scatter(centers[:, 0], centers[:, 1],
              marker='o', c="white", alpha=1, s=200)

  for i, c in enumerate(centers):
      ax2.scatter(c[0], c[1], marker='$%d$' % i, alpha=1, s=50)

  ax2.set_title("The visualization of the clustered data.")
  ax2.set_xlabel("Feature space for the 1st feature")
  ax2.set_ylabel("Feature space for the 2nd feature")

  plt.suptitle(("Silhouette analysis for KMeans clustering on sample data "
                "with n_clusters = %d" % n_clusters),
               fontsize=14, fontweight='bold')

  plt.show()


#Based on above findings choosing k=10

kmeans_model = KMeans(n_clusters=10, random_state=1234)

kmeans_model.fit(X1)

clusters = kmeans_model.labels_

#Using kmodes 

from kmodes.kmodes import KModes

km = KModes(n_clusters=10, init='Huang', n_init=5, verbose=1)

clusters = km.fit_predict(X1)

#Using Kprototype
from kmodes.kprototypes import KPrototypes

kproto = KPrototypes(n_clusters=n_clusters, init='Huang', verbose=1)

#Don't execute only with numerical variables
clusters = kproto.fit_predict(X1) #Use ,categorical=[1, 2])

#Market Basket analysis

from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

productdata = segdata[['InvoiceNo','Description']]

productdata['values'] = 1

prod_data = pd.crosstab(index = segdata['InvoiceNo'],columns =segdata['Description'])

frequent_itemsets = apriori(prod_data, min_support=0.001, use_colnames=True,max_len = 3)

rules = association_rules(frequent_itemsets, metric="lift", min_threshold=1)

rules.head()
